### PR TITLE
fix microsoft auth and ingestion for users without defined email

### PR DESCRIPTION
**Pull request by JessicaJHee**
## Hey, I just made a Pull Request!
This PR is to address the issue with users not being ingested into the catalog when they don't have an email defined in their profile.
* `metadata.name` is set to the `user.userPrincipalName` when `user.mail` is undefined
* the email annotation (`user.metadata.annotations[MICROSOFT_EMAIL_ANNOTATION]`) and user spec (`user.spec.profile.email`) is left blank when `user.mail` is undefined

To address the concern with the sign in resolver using email to match, I've added a new resolver that uses the user Id to do the matching to fall back on when the email field does not exist. Please let me know how this looks and I'll update the documentation if this is the approach we want to go with.

Fixes https://github.com/backstage/backstage/issues/26137

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by freben** [comment]
Check the build output - you'll want to run `yarn build:api-reports` whenever interfaces change, and commit those changes with your PR.

Also to ensure that changes are actually shipped, you want to make changesets for the modified packages. See the bot comment above. Run`yarn changeset` and follow the instructions.

**Comment by JessicaJHee** [comment]
cc: @awanlin @afscrome

File: .changeset/real-pants-rule.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/plugin-auth-backend-module-microsoft-provider': minor
```
Comment (freben) [inline .changeset/real-pants-rule.md:2]:
```suggestion
'@backstage/plugin-auth-backend-module-microsoft-provider': patch
'@backstage/plugin-auth-backend': patch
```

it's not really breaking as far as I can tell.

**Review by freben — APPROVED** [review]

**Comment by awanlin** [comment]
Hi @JessicaJHee, this got merged faster than I could add a review. Can you create a follow up PR adding details about this new resolve to the documentation here, please? https://backstage.io/docs/auth/microsoft/provider#resolvers

**Pull request merged**